File: killstreek/src/main/java/dev/kkorolyov/killstreek/item/Weapon.java
package dev.kkorolyov.killstreek.item;

import dev.kkorolyov.killstreek.utility.BoundedValue;
import dev.kkorolyov.pancake.component.Sprite;

/**
 * An offensive equippable.
 */
public class Weapon extends Equippable {
	private final float rate;
	private final int requiredSlots;

	/**
	 * Constructs a new weapon.
	 * @param name weapon name
	 * @param sprite weapon visual
	 * @param durability weapon durability
	 * @param value damage value
	 * @param rate weapon fire rate in s
	 * @param requiredSlots number of weapon slots required to equip this weapon
	 */
	public Weapon(String name, Sprite sprite, BoundedValue<Integer> durability, int value, int rate, int requiredSlots) {
		super(name, sprite, durability, value);

		this.rate = rate;
		this.requiredSlots = requiredSlots;
	}

	/** @return fire rate in s */
	public float getRate() {
		return rate;
	}

	/** @return number of weapon slots required to equip this weapon */
	public int getRequiredSlots() {
		return requiredSlots;
	}
}

File: killstreek/src/main/kotlin/dev/kkorolyov/killstreek/item/Weapon.kt
package dev.kkorolyov.killstreek.item

import dev.kkorolyov.killstreek.media.Sprite
import dev.kkorolyov.pancake.platform.math.BoundedValue

/**
 * An offensive [Equippable],
 * @constructor constructs a new weapon with attack [rate] and [requiredSlots] to equip
 */
class Weapon(
		name: String,
		sprite: Sprite,
		durability: BoundedValue<Int>,
		value: Int,
		/** attack rate in `ns` */
		val rate: Int,
		/** number of weapon slots required to equip this weapon */
		val requiredSlots: Int
) : Equippable(name, sprite, durability, value)
